### PR TITLE
fix(hydroflow): cleanup temp tcp networking code, fix race condition fix #1458

### LINKED ISSUE #1458: Race condition in TCP & WebSockets Stack
The [bind_tcp](https://github.com/hydro-project/hydroflow/blob/afe78c343658472513b34d28658634b253148aee/hydroflow/src/util/tcp.rs#L96) method has a race condition. The client is removed from the hashmap and and back to it across an `await()` boundary. This leaves the client map inconsistent when execution awaits at the await point (the client exists but cannot be found in the map).

Since the newly introduced WebSockets examples rely on the same patterns as TCP, they also have the same race condition.

### PR DESCRIPTION
**Pull request by MingweiSamuel**
consolidate into one task to prevent races

### CODE REVIEW

File: hydroflow/src/util/tcp.rs [hunk 1]
```diff
@@ -74,6 +73,7 @@ pub type TcpFramedSink<T> = Sender<(T, SocketAddr)>;
 pub type TcpFramedStream<Codec: Decoder> =
     Receiver<Result<(<Codec as Decoder>::Item, SocketAddr), <Codec as Decoder>::Error>>;
 
+// TODO(mingwei): this temporary code should be replaced with a properly thought out networking system.
```
Comment (rohitkulshreshtha) [inline hydroflow/src/util/tcp.rs:76]: Indeed.

File: hydroflow/src/util/tcp.rs [hunk 2]
```diff
@@ -83,60 +83,63 @@ pub async fn bind_tcp<T: 'static, Codec: 'static + Clone + Decoder + Encoder<T>>
...
+
         loop {
-            let (stream, peer_addr) = if let Ok((stream, _)) = listener.accept().await {
-                if let Ok(peer_addr) = stream.peer_addr() {
-                    (stream, peer_addr)
-                } else {
-                    continue;
+            // Calling methods in a loop, futures must be cancel-safe.
+            select! {
+                biased;
```
Comment (rohitkulshreshtha) [inline hydroflow/src/util/tcp.rs:97]: It will be helpful to document why `biased` is necessary.

File: hydroflow/src/util/tcp.rs [hunk 3]
```diff
@@ -83,60 +83,63 @@ pub async fn bind_tcp<T: 'static, Codec: 'static + Clone + Decoder + Encoder<T>>
...
-                let clients = clients.clone();
-                async move {
-                    let mapped = recv.map(|x| Ok(x.map(|x| (x, peer_addr))));
-                    let _ = tx_ingress.send_all(&mut pin!(mapped)).await;
-
-                    clients.borrow_mut().remove(&peer_addr);
+                // Send outgoing messages.
+                msg_send = recv_egress.next() => {
+                    let Some((payload, peer_addr)) = msg_send else {
+                        continue;
```
Comment (rohitkulshreshtha) [inline hydroflow/src/util/tcp.rs:119]: In what condition is this `continue` called?

File: hydroflow/src/util/tcp.rs [hunk 4]
```diff
@@ -83,60 +83,63 @@ pub async fn bind_tcp<T: 'static, Codec: 'static + Clone + Decoder + Encoder<T>>
...
-                let client = clients.borrow_mut().remove(&addr);
-
-                if let Some(mut sender) = client {
-                    let _ = SinkExt::send(&mut sender, payload).await;
-                    clients.borrow_mut().insert(addr, sender);
-                }
-            }
-        }
-    });
+    let (send_egress, mut recv_egress) = unsync_channel::<(T, SocketAddr)>(None);
```
Comment (rohitkulshreshtha) [inline hydroflow/src/util/tcp.rs:86]: Curious, what was wrong with `tx` and `rx`?

File: hydroflow/src/util/tcp.rs [hunk 5]
```diff
@@ -147,34 +150,53 @@ pub async fn bind_tcp<T: 'static, Codec: 'static + Clone + Decoder + Encoder<T>>
...
-                    spawn_local(async move {
-                        let mapped = recv.map(|x| Ok(x.map(|x| (x, addr))));
-                        let _ = tx_ingress.send_all(&mut pin!(mapped)).await;
-                    });
-
-                    entry.insert(send)
+        loop {
+            // Calling methods in a loop, futures must be cancel-safe.
+            select! {
+                biased;
```
Comment (rohitkulshreshtha) [inline hydroflow/src/util/tcp.rs:164]: Ditto on why `biased` is necessary.

File: hydroflow/src/util/tcp.rs (hunk 4 above)
Comment (MingweiSamuel) [reply hydroflow/src/util/tcp.rs:86]: I just prefer the more explicit naming, consistent with how rust names channels

File: hydroflow/src/util/tcp.rs (hunk 2 above)
Comment (rohitkulshreshtha) [reply hydroflow/src/util/tcp.rs:98]: What happens if you don't put `biased`?

Comment (rohitkulshreshtha) [reply hydroflow/src/util/tcp.rs:98]: I think I forgot to unresolve this conversation.

Explaining what happens if `biased` is removed from here will be helpful. It seems to be one word: doing a lot of heavy lifting.

Comment (jhellerstein) [reply hydroflow/src/util/tcp.rs:98]: I have no idea what biased means :-)

File: hydroflow/src/util/tcp.rs [hunk 6]
```diff
@@ -83,60 +83,65 @@ pub async fn bind_tcp<T: 'static, Codec: 'static + Clone + Decoder + Encoder<T>>
...
+                new_peer = listener.accept() => {
+                    let Ok((stream, _addr)) = new_peer else {
+                        continue;
+                    };
+                    let Ok(peer_addr) = stream.peer_addr() else {
+                        continue;
+                    };
+                    let (peer_send, peer_recv) = tcp_framed(stream, codec.clone());
+
+                    // TODO: Using peer_addr here as the key is a little bit sketchy.
```
Comment (jhellerstein) [inline hydroflow/src/util/tcp.rs:109]: I think this merits filing an issue to be resolved in future.

**Review by jhellerstein — APPROVED** [review]
Overarching comment is that this code is necessarily very case-based, but it's hard on reading it to understand the space of possibilities and whether they're covered. It would be really nice if this was better documented -- or refactored as a state machine or something easier to reason about.

This relates to tests as well -- I'm not seeing any here, and don't know how you propose to cover the cases in tests.

I'm not opposed to merging the PR, but if so we should file some cleanup issues for the future?

File: hydroflow/src/util/tcp.rs (hunk 6 above)
Comment (MingweiSamuel) [reply hydroflow/src/util/tcp.rs:109]: #1506

File: hydroflow/src/util/tcp.rs (hunk 2 above)
Comment (MingweiSamuel) [reply hydroflow/src/util/tcp.rs:104]: Added more detailed comments for your reading pleasure :P

**Comment by MingweiSamuel** [comment]
b96123369cec3d6d407af973f1896b2734fd92ef

**Pull request merged**